File: src/PresenterAi.Application/Presenting/EndReasons.cs
namespace PresenterAi.Application.Presenting;

public static class EndReasons
{
    public const string User = "user";
    public const string Completed = "completed";
    public const string MaxLength = "max_length";
    public const string Idle = "idle";
    public const string Heartbeat = "heartbeat";
    public const string WriterFailed = "writer_failed";
    public const string Backpressure = "backpressure";
    public const string Disconnect = "disconnect";
    public const string Takeover = "takeover";
    public const string Shutdown = "shutdown";
    public const string Error = "error";
    public const string UpstreamLost = "upstream_lost";
    public const string ReconnectFailed = "reconnect_failed";
    public const string CliCancelled = "cli_cancelled";
    public const string CliMaxSeconds = "cli_max_seconds";
    public const string StopAfterSlide = "stop_after_slide";

    public static readonly IReadOnlySet<string> All = new HashSet<string>(StringComparer.Ordinal)
    {
        User,
        Completed,
        MaxLength,
        Idle,
        Heartbeat,
        WriterFailed,
        Backpressure,
        Disconnect,
        Takeover,
        Shutdown,
        Error,
        UpstreamLost,
        ReconnectFailed,
        CliCancelled,
        CliMaxSeconds,
        StopAfterSlide
    };
}

File: web/app/src/utils/endReasons.ts
export const END_REASON_TEXT: Record<string, string> = {
  user: "Talk ended by user",
  completed: "Talk completed",
  max_length: "Maximum talk length reached",
  idle: "Ended due to inactivity",
  heartbeat: "Connection lost (heartbeat timeout)",
  writer_failed: "Connection failed",
  backpressure: "Connection closed due to backpressure",
  disconnect: "Disconnected",
  takeover: "Taken over in another tab",
  shutdown: "Server shutting down",
  error: "An error occurred",
  upstream_lost: "Upstream connection lost",
  reconnect_failed: "Failed to reconnect",
  cli_cancelled: "Cancelled from CLI",
  cli_max_seconds: "Maximum duration reached",
  stop_after_slide: "Stopped after slide",
};

export function formatEndReason(endReason: string | null | undefined): string | null {
  if (!endReason) return null;
  return END_REASON_TEXT[endReason] ?? endReason;
}
